frontend-react/docs/test-conventions.md

```diff
@@ -0,0 +1,492 @@
+    test("returns true if the value contains a number", () => {
+        expect(containsNumber("123")).toEqual(true);
+        expect(containsNumber("abc123")).toEqual(true);
+        expect(containsNumber("123abc")).toEqual(true);
+        ...
+    });
+    test("returns false if the value does not contain a number", () => {
+        expect(containsNumber("abc")).toEqual(false);
+        expect(containsNumber("abcxyz")).toEqual(false);
+        ...
+    });   
+});
+```
+
+> Why?
+>
+>Type-checking can of course help mitigate pessimistic cases since disallowed types will throw errors at compilation time, but any application with real users should account for unexpected values and logical blind spots.  This is especially important for optional function arguments and object properties due to the possibility of falsy values.
+
+### Tests should validate _custom logic_, not third-party logic
+
+If a component will always render in the same way given its props, there's no point in testing that because that's fundamentally testing React's renderer.  However, if there's custom forking logic in how a component renders, that's _our_ logic so it should be tested.
+
+```tsx
+// bad - this component will always render in the same way
+// so we're ultimately just testing that React works properly
+function StaticComponent({ message }: { message: string }) {
+    return (
+        <div>
+            <p>{ message }</p>
+        </div>
+    );
+}
+
+describe("StaticComponent", () => {
+    beforeEach(() => {
+        render(<StaticComponent message="Hello!" />);
+    });
+
+    test("renders correctly", () => {
+        expect(screen.getByText("Hello!")).toBeVisible();
+    });
+});
+
+// good - we have forking logic so we should verify it works
+function DynamicComponent({ message }: { message?: string }) {
+    if (message) {
+        return <p>{message}</p>;
+    }
+    
+    return <p>This is a fallback message.</p>;
+}
+
+describe("DynamicComponent", () => {
+    describe("when no message is provided", () => {
+        beforeEach(() => {
+            render(<DynamicComponent />);
+        });
+
+        test("renders the fallback message", () => {
+            expect(screen.getByText("This is a fallback message.")).toBeVisible();
+        });
+    });
+
+    describe("when a message is provided", () => {
+        beforeEach(() => {
+            render(<DynamicComponent message="Hello!" />);
+        });
+
+        test("renders the provided message", () => {
+            expect(screen.getByText("Hello!")).toBeVisible();
+        });
+        
+        test("does not render the fallback message", () => {
+            expect(screen.queryByText("This is a fallback message.")).toBeNull();
+        });    
+    });
+});
+```
+
+> Why?
+>
+> It's redundant to test third-party library logic, even implicitly, because that logic is entirely out of our control and really should already be tested by the maintainers of said library.
```

Comment:
Woops, looks like you cover this here kinda in regards to my react-query example.

Reply:
No problem!  Happy to update this for something a bit more practical with `react-query` -- do you have that example ready?  Or is it just something like the following?

```tsx
export function useSomething() {
    const { isLoading } = useQuery(...);

    return {
        isLoading
    };
}

test(useSomethingReturnValue.isLoading).toEqual(true); // redundant to test react-query logic
```

Reply:
Pretty much that exactly. We shouldn't test third party logic, but it's OK to rely on it to know when to call an expect (waitFor isLoading to be false).

Reply:
Makes sense to me!  I'll add that as an example too -- thanks!